### DISCUSSION TITLE
Running Firefly III in a docker container under a subdirectory using Apache

### DISCUSSION THREAD

**Discussion by cartbar** [firefly-iii/firefly-iii#4892] [General]
I have seen a number of comments about how to get Firefly III running in a Docker container, but not under the root directory.  This is my solution.  

First I installed a standalone MariaDB database (i.e. not in a Docker container - not strictly revelvant)

The Firefly docker container was created using

`
sudo docker container create -v firefly_iii_upload:/var/www/html/storage/upload -p8087:8080 -e APP_KEY=... -e DB_HOST=172.17.0.1 -e DB_PORT=3306 -e DB_CONNECTION=mysql -e DB_DATABASE=firefly_iii -e DB_USERNAME=... -e DB_PASSWORD=... -e TRUSTED_PROXIES="**" --network bridge --restart unless-stopped --name FireflyIII fireflyiii/core:latest
`

I installed a standalone Apache server, to act as an SSL terminator (i.e. it handles the SSL so that the Firefly server doesn't have to.

In the Apache server config I put:

```
<IfModule mod_ssl.c>
<VirtualHost *:443>
        ServerAdmin webmaster@localhost
        DocumentRoot /var/www/html

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        ServerName ...
        SSLCertificateFile /etc/letsencrypt/live/.../fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/.../privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf

        <Location /accounts>
            ProxyAddHeaders Off
            RequestHeader unset Accept-Encoding
            ProxyPass http://localhost:8087
            ProxyPassReverse http://localhost:8087
            AddOutputFilterByType SUBSTITUTE text/html
            Substitute "s|http://localhost:8087/|https://.../accounts/|i"
        </Location>
</VirtualHost>
</IfModule>
```

ProxyAddHeaders is off to prevent Apache sending the X-Forward headers to Firefly, otherwise the URLs it populates in the response will include the correct website, but not the correct path.

Unsetting the Accept-Encoding header ensures that the response from Firefly is not compressed, otherwise the "Substitute" directive won't work

ProxyPass and ProxyPassReverse act as you would expect, so that Firefly will appear under the "accounts" directory.

The AddOutputFilterByType and Substitute directives alter the pages returned from Firefly, replacing the localhost urls with the website's url (obviously replace "..." with your website address

So far, it works as expected, so I hope this helps someone else

**Comment by JC5** [comment]
Cool, thanks for sharing!

**Comment by Galang23** [comment]
This is what I was looking for the past 30 days. It's just beautiful and useful. Thank you for sharing this precious information! Maintainer might add these in FAQ section or pin it somewhere to inform new users.

**Comment by JC5** [comment]
It is now in the FAQ!

**Comment by bencarp** [comment]
I couldn't find the nginx equivalent anywhere, so here it is, I hope it helps someone:
```
server {
    server_name example.com;
    root /var/www/html;

    listen [::]:443 ssl;
    listen 443 ssl;
    ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem; 
    ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem;

    location /accounts/ {

                proxy_pass http://localhost:8087/; # Trailing slash is IMPORTANT, it won't work without
                proxy_redirect "http://localhost:8087" "https://example.com/accounts"; # Replace "Location" response header (Login, Logout)
                proxy_set_header Accept-Encoding ""; # Tell Firefly not to compress response
                sub_filter "http://localhost:8087" "https://example.com/accounts";  # Find and replace URL in response body
                sub_filter_once off; # Replace every instance

                proxy_read_timeout 10s; # Optional, helps with debugging if something's wrong
                proxy_send_timeout 10s; # Same here

    }
}

**Reply by bencarp** [reply]
This doesn't yet work 100% on the data-importer (same configuration, data-importer also under a separate subdirectory). When being redirected back from Nordigen, it always tries to send me to localhost. For now, I don't quite understand why @JC5

**Reply by lits101** [reply]
> I couldn't find the nginx equivalent anywhere, so here it is, I hope it helps someone:
> 
> ```
> server {
>     server_name example.com;
>     root /var/www/html;
> 
>     listen [::]:443 ssl;
>     listen 443 ssl;
>     ssl_certificate /etc/letsencrypt/live/example.com/fullchain.pem; 
>     ssl_certificate_key /etc/letsencrypt/live/example.com/privkey.pem;
> 
>     location /accounts/ {
> 
>                 proxy_pass http://localhost:8087/; # Trailing slash is IMPORTANT, it won't work without
>                 proxy_redirect "http://localhost:8087" "https://example.com/accounts"; # Replace "Location" response header (Login, Logout)
>                 proxy_set_header Accept-Encoding ""; # Tell Firefly not to compress response
>                 sub_filter "http://localhost:8087" "https://example.com/accounts";  # Find and replace URL in response body
>                 sub_filter_once off; # Replace every instance
> 
>                 proxy_read_timeout 10s; # Optional, helps with debugging if something's wrong
>                 proxy_send_timeout 10s; # Same here
> 
>     }
> }
> ```

This works great! I'm not using the data importer though.

**Reply by AntarEspadas** [reply]
In addition to this, I had to add `sub_filter_types *;` for things like date ranges and bulk delete to work. This applies substitutions to all types of files, including JS files, which may have unforeseen consequences, but seems to be working fine for now.

**Reply by angelcervera** [reply]
@AntarEspadas Could you share the final configuration?

**Reply by Cloud9Developer** [reply]
This was almost exactly what was needed! Had one issue when using my custom DNS (sub pages wouldn't load/returned 404). Use the below code and it should work!

The only thing still kinda broken is the auth page. You can access the auth page (following the path set below) like: `https://example.com/budget/login` and login, but the redirect after logging in doesn't work. However, the cookie does save and if you go to the home page after logging in, it works fine :smile: 

```nginx
server {
        server_name example.com;
        listen 443 ssl http2;
        listen [::]:443 ssl http2;      
        include snippets/self-signed.conf; # I have my certs here
        include snippets/custom-ssl-params.conf; # My common ssl params
        include snippets/proxy-params.conf; # Common Proxy Params

# Set up the context root "/budget"
    location /budget/ {
        proxy_pass http://localhost:8087/;
        proxy_redirect "http://localhost:8087" "https://$host$uri";
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        # Append URI paths to the request
        #internal;
        sub_filter "http://localhost" "https://$host"; # This will only work (at least for me) if you're connecting from localhost
        sub_filter "https://$host" "https://$host/budget"; # This will allow it to work if you have a custom DNS entry setup
        sub_filter_once off;
        sub_filter_types *;
        proxy_pass_request_headers on;
    }
}
```

**Comment by toffee** [comment]
@cartbar thanks this solution. 

In my environment the "Dashboard" still points to http://dockerhostname:dockerport.
I've fix this by changing the line:
` Substitute "s|http://localhost:8087/|https://.../accounts/|i"`
with (notice the missing trailing slash)
` Substitute "s|http://localhost:8087|https://.../accounts|i"`

Anyway, seems that this does not work for all URLs (I guess the ones on client side). If I try to change the daterage (in top right of the page) nothing happens because browser console display error `Could not change date range`
This comes from [firefly.js](https://github.com/firefly-iii/firefly-iii/blob/4804f419ef158077bd5c55b9c81f5a2d58ab611f/public/v1/js/ff/firefly.js#L90) because the URL (`dateRangeMeta.url`) is invalid (point to the http://dockerhostname:dockerport)

How this can be fixed?

**Reply by AntarEspadas** [reply]
I faced a similar issue with nginx. It was due to substitutions only being applied to certain MIME types, like HTML, and not JS. Maybe [this SO thread](https://stackoverflow.com/questions/41661153/reverseproxy-how-to-change-content-in-embedded-javascript-filel) can help you?
